nixos/doc/manual/release-notes/rl-2105.xml

```diff
@@ -662,6 +662,24 @@ self: super:
      The option's description was incorrect regarding ownership management and has been simplified greatly.
     </para>
    </listitem>
+   <listitem>
+    <para>
+     The <literal>security.apparmor</literal> module,
+     for the <link xlink:href="https://gitlab.com/apparmor/apparmor/-/wikis/Documentation">AppArmor</link>
+     Mandatory Access Control system,
+     has been substantialy improved along with related tools,
+     so that module maintainers can now more easily write AppArmor profiles for NixOS.
+     The most notable change on the user-side is the new option <xref linkend="opt-security.apparmor.policies"/>,
+     replacing the previous <literal>profiles</literal> option
+     to provide a way to disable a profile
+     and to select whether to confine in enforce mode (default)
+     or in complain mode (see <literal>journalctl -b --grep apparmor</literal>).
+     Before enabling this module, either directly
+     or by importing <literal>&lt;nixpkgs/nixos/modules/profiles/hardened.nix&gt;</literal>,
```

Comment:
Thank you for continuing this work. However, I think I didn't realize this last year — it's not merely people knowingly enabling AppArmor for the first time but all users of the hardened profile¹ who, upon updating (perhaps automatically) to a version of nixpkgs that includes these changes, will get any already-running confinable processes killed? What's the risk of data loss, especially with `system.autoUpgrade`?

<small>¹ unless they've specifically disabled AppArmor for some reason</small>

Reply:
@8573, yes, the `hardened` profile is enabling `AppArmor`, hence people importing it would have their unconfined and newly confinable processes killed by default. Because `AppArmor` cannot confine them without killing them, enabling `killUnconfinedConfinables` by default is indeed privileging security over stability. I would argue that it respects the expectation of users importing a profile named: `hardened`. It would be more problematic if `AppArmor` were to be enabled by default on NixOS. So I guess the sensible thing is to disable `killUnconfinedConfinables` by default but enable it in the `hardened` profile, so that it can be enabled conciously by users privileging security, while privileging stability for all other users in case `AppArmor` becomes enabled by default one day.

Concerning data loss, `killUnconfinedConfinables` uses `kill(1)` to send a `SIGTERM`, not a `SIGKILL`, hence the processes have a chance to properly terminate without data loss or corruption, but not all are coded to graciously terminate like so. Ideally, they should also be restarted by `systemd`, but the service can be configured otherwise, and this is only for processes supervised by `systemd` or an other mecanism.

Concerning `system.autoUpgrade`, the introduction of a new (enabled by default) `AppArmor` profile in NixOS will indeed try to kill the processes matching that profile.